lib/api/src/rc_switch_user.cpp

```diff
@@ -0,0 +1,59 @@
+#include "irods/switch_user.h"
+
+#include "irods/plugins/api/switch_user_types.h"
+#include "irods/plugins/api/api_plugin_number.h"
+#include "irods/procApiRequest.h"
+#include "irods/rodsErrorTable.h"
+
+#include <cstdlib>
+#include <cstring>
+
+auto rc_switch_user(RcComm* _comm, const char* _username, const char* _zone) -> int
+{
+    // NOLINTNEXTLINE(readability-implicit-bool-conversion)
+    if (!_username || !_zone) {
+        return SYS_INVALID_INPUT_PARAM;
+    }
+
+    auto& client = _comm->clientUser;
+
+    // Attempting to verify that the user is already the target user cannot be done without
+    // going over the wire. That's because the client cannot determine the user type of the
+    // target user. For that reason, this implementation always results in a network request.
+
+    try {
+        SwitchUserInput input{};
+        std::strncpy(static_cast<char*>(input.username), _username, sizeof(SwitchUserInput::username));
+        std::strncpy(static_cast<char*>(input.zone), _zone, sizeof(SwitchUserInput::zone));
+
+        SwitchUserOutput* output{};
+        void** out_ptr = reinterpret_cast<void**>(&output); // NOLINT(cppcoreguidelines-pro-type-reinterpret-cast)
+
+        const int ec = procApiRequest(_comm, SWITCH_USER_APN, &input, nullptr, out_ptr, nullptr);
+
+        // NOLINTNEXTLINE(readability-implicit-bool-conversion)
+        if (ec == 0) {
+            // On success, we always assume that the output pointer is valid.
+            // If the server-side implementation is solid, then we shouldn't need to check the pointer.
+
+            // NOLINTBEGIN(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+            std::strncpy(client.userName, _username, sizeof(UserInfo::userName));
+            std::strncpy(client.rodsZone, _zone, sizeof(UserInfo::rodsZone));
+            std::strncpy(client.userType, output->user_type, sizeof(UserInfo::userType));
```

Comment:
Investigate whether the client connection requires `userType`.

If it isn't set following `rcConnect()` or `clientLogin()` then it is probably unused on the client-side. That also means we CAN short-ciruit this API call.

Reply:
Done, but we need to be sure we are happy with the changes.